src/Reflex/Dynamic/FactorDynGeneric.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
+module Reflex.Dynamic.FactorDynGeneric
+  (
+    factorDyn
+  , factorDynGeneric
+  ) where
+
+import Control.Monad.Fix (MonadFix)
+import Data.Dependent.Sum (DSum ((:=>)))
+import Data.Functor.Compose (Compose (Compose), getCompose)
+import Data.GADT.Compare ((:~:) (Refl), GEq, geq)
+import Data.Maybe (isJust)
+import Reflex (Dynamic, Event, MonadHold, Reflex, current, ffor, fforMaybe, fmapMaybe, hold, holdDyn, never,
+               push, sample, switch, updated)
+
+import Generics.SOP ((:.:) (Comp), Code, Generic, I (I), Proxy (Proxy), SListI, SOP (SOP), from, hcmap, hmap,
+                     to, unSOP)
+import Generics.SOP.Distribute (distributeI_NP)
+import Generics.SOP.DMapUtilities (FunctorWrapTypeListOfLists, dSumToNS, nsOfnpUnCompose, nsToDSum)
+
+-- | Use factorDyn and generics-sop to turn a Dynamic of a type into a factored Dynamic of a type with Dynamic fields.  See below functions for examples
+factorDynGeneric :: forall t m a b. (Reflex t, MonadFix m, MonadHold t m
+                                    , Generic a, Generic b
+                                    , (Code b) ~ FunctorWrapTypeListOfLists (Dynamic t) (Code a))
+  => Dynamic t a -> m (Dynamic t b)
+factorDynGeneric da = do
+  let dSumDyn = nsToDSum . unSOP . from <$> da -- Dynamic t (DSum (TypeListTag (Code a)) (NP I xs))
+  dSumDyn' <- factorDyn dSumDyn  -- Dynamic t (DSum (TypeListTag (Code a)) (NP I))
+  let nsnpDyn = dSumToNS <$> dSumDyn' -- Dynamic t (NS (Compose (Dynamic t) (NP I)) (Code a))
+      sListIC = Proxy :: Proxy SListI
+      nsnpDyn' = hcmap sListIC (hmap (Comp. I) . distributeI_NP . getCompose) <$> nsnpDyn -- Dynamic t (NS (NP (I :.: Dynamic t)) (Code a))
+      nsnpDyn'' = nsOfnpUnCompose <$> nsnpDyn' -- Dynamic t (NS (NP I) (Code b))
+      result = to . SOP <$> nsnpDyn''
+  return result
+
+-- for example:
+maybeDyn :: (Reflex t, MonadHold t m, MonadFix m) => Dynamic t (Maybe a) -> m (Dynamic t (Maybe (Dynamic t a)))
+maybeDyn = factorDynGeneric
+
+eitherDyn :: (Reflex t, MonadHold t m, MonadFix m) => Dynamic t (Either a b) -> m (Dynamic t (Either (Dynamic t a) (Dynamic t b)))
```

Comment:
This is really cool!

Reply:
Thanks!  I still think there may be a bug but though I first found it a while ago, I haven't had a chance to go back and update the reflex version (maybe I'll try that now) or make a simpler case (I found it amidst all the profunctor stuff).  But we should definitely make sure there are a few tests.